packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs

```diff
@@ -1022,4 +1047,156 @@ where
+            .query(
+                r#"
+                WITH aggregated as (
+                    SELECT link_type_version_id, ARRAY_AGG(target_entity_id ORDER BY multi_order ASC) as links
+                    FROM links
+                    WHERE active AND multi and source_entity_id = $1
+                    GROUP BY link_type_version_id
+                )
+                SELECT base_uri, "version", links FROM aggregated
+                INNER JOIN ids ON ids.version_id = aggregated.link_type_version_id
+                "#,
+                &[&source_entity_id],
+            )
+            .await
+            .report()
+            .change_context(QueryError)
+            .attach_printable(source_entity_id)?
+            .into_iter()
+            .map(|row| (VersionedUri::new(row.get(0), row.get::<_, i64>(1) as u32), OutgoingLink::Multiple(row.get(2))));
+
+        let single_links = self
+            .client
+            .as_client()
+            .query(
+                r#"
+                SELECT base_uri, "version", target_entity_id
+                FROM links
+                INNER JOIN ids ON ids.version_id = links.link_type_version_id
+                WHERE active AND NOT multi and source_entity_id = $1
+                "#,
+                &[&source_entity_id],
+            )
+            .await
+            .report()
+            .change_context(QueryError)
+            .attach_printable(source_entity_id)?
+            .into_iter()
+            .map(|row| {
+                (
+                    VersionedUri::new(row.get(0), row.get::<_, i64>(1) as u32),
+                    OutgoingLink::Single(row.get(2)),
+                )
+            });
+
+        Ok(Links::new(multi_links.chain(single_links).collect()))
+    }
+
+    async fn get_link_target(
+        &self,
+        source_entity_id: EntityId,
+        link_type_uri: VersionedUri,
+    ) -> Result<OutgoingLink, QueryError> {
+        let version = i64::from(link_type_uri.version());
+        let links = self
+            .client
+            .as_client()
+            .query(
+                r#"
+                SELECT target_entity_id
+                FROM links
+                INNER JOIN ids ON ids.version_id = links.link_type_version_id
+                WHERE active AND source_entity_id = $1 AND base_uri = $2 AND "version" = $3
+                "#,
+                &[&source_entity_id, link_type_uri.base_uri(), &version],
+            )
+            .await
+            .report()
+            .change_context(QueryError)
+            .attach_printable(source_entity_id)?;
+
+        if links.is_empty() {
+            return Err(Report::new(QueryError)
+                .attach_printable(source_entity_id)
+                .attach_printable(link_type_uri.clone()));
+        } else if links.len() == 1 {
+            return Ok(OutgoingLink::Single(
+                links
+                    .get(0)
+                    .expect("Link array should have at least one element")
```

Comment:
I don't like this a lot. Any better way to conditionally do something when the vec is only length one?